internal/controllers/nodefeaturerule_reconciler.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2021. The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package new_controllers
+
+import "C"
+import (
+	"context"
+	"github.com/google/go-cmp/cmp"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	nfdk8ssigsiov1alpha1 "github.com/openshift/cluster-nfd-operator/api/v10"
+	nfdopenshiftiov1alpha1 "github.com/openshift/cluster-nfd-operator/api/v1alpha1"
+)
+
+// NodeFeatureRuleReconciler reconciles a NodeFeatureRule object
+type nodeFeatureRuleReconciler struct {
+	client client.Client
+}
+
+func NewNodeFeatureRuleReconciler(client client.Client) *nodeFeatureRuleReconciler {
+	return &nodeFeatureRuleReconciler{
+		client: client,
+	}
+}
+
+// +kubebuilder:rbac:groups=nfd.openshift.io,resources=nodefeaturerules,verbs=get;list;watch
+// +kubebuilder:rbac:groups=nfd.k8s-sigs.io,resources=nodefeaturerules,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// the NodeFeatureRule object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+func (r *nodeFeatureRuleReconciler) Reconcile(ctx context.Context, nfr *nfdopenshiftiov1alpha1.NodeFeatureRule) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	logger.Info("Reconciling NodeFeatureRule from nfd.openshift.io group", "name", nfr.Name)
+
+	target := &nfdk8ssigsiov1alpha1.NodeFeatureRule{}
+	err := r.client.Get(ctx, client.ObjectKey{Name: nfr.Name, Namespace: nfr.Namespace}, target)
```

Comment:
I think we can improve that code with a couple of things:
1) lets check the GetDeletionTimestamp field of the nfr: if it is set, it means that the openshift NFR is being deleted, so no need to to do anything
2) we can user CreateOrPatch function: it will create the object if it does not exists, and in case it exists will compare the target with the existing one and update it only in case there is any change